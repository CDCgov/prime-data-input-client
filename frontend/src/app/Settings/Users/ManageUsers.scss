.users-sidenav {
  border-right: 1px solid #dfe1e2; // color("text-base-lighter");
}

.users-sidenav-item:hover {
  cursor: pointer;
}
.user-header {
  border-bottom: 1px solid #dfe1e2;
  padding: 0em 1em;
}

.user-content {
  padding: 0em 1em;
}

.reset-button {
  display: block;
  margin: 1em auto;
}

.usa-current {
  .sidenav-user-name {
    color: #005ea2;
  }
}

.sidenav-user-name {
  color: #171717; // color($theme-color-prime-gray-light)
}

.sidenav-user-email {
  font-weight: normal;
  font-size: 0.8em;
}

.facility-list {
  td {
    border: 1px solid #dfe1e2;
  }
}

.remove-tag {
  cursor: pointer;
}

.close-button {
  display: block;
  border: none;
  background: transparent;
  cursor: pointer;
}
