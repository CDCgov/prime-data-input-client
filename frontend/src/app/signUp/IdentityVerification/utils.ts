import * as yup from "yup";

const EXPERIAN_ANSWER_KEY = "outWalletAnswer";

export const getAnswerKey = (index: number) =>
  `${EXPERIAN_ANSWER_KEY}${index + 1}`;

export const toOptions = (
  choices: string[]
): { value: string; label: string }[] =>
  choices.map((choice, index) => {
    return { value: String(index + 1), label: choice };
  });

export const initAnswers = (questionSet: Question[]): Nullable<Answers> =>
  questionSet.reduce((answers, _question, index) => {
    answers[getAnswerKey(index)] = null;
    return answers;
  }, {} as Nullable<Answers>);

export const buildSchema = (questionSet: Question[]): yup.SchemaOf<Answers> =>
  yup.object(
    questionSet.reduce((answers, _question, index) => {
      answers[getAnswerKey(index)] = yup
        .string()
        .nullable()
        .required("This field is required");
      return answers;
    }, {} as { [key: string]: any })
  );
