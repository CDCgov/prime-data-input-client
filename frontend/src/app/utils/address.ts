export function formatAddress(address: Address) {
  // format address with correct punctuation and separated by `\n`s
  const lastAddressLine = `${address.city}${
    address.state && address.city ? ", " : ""
  }${address.city && address.zipCode && !address.state ? " " : ""}${
    address.state
  }${address.state && address.zipCode ? " " : ""}${address.zipCode}`;

  let result = address.street;
  result += `${
    result.length > 0 && address.streetTwo
      ? "\n" + address.streetTwo
      : address.streetTwo
  }`;
  result += `${
    result.length > 0 && lastAddressLine
      ? "\n" + lastAddressLine
      : lastAddressLine
  }`;
  return result;
}
