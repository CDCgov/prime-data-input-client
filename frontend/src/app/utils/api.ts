interface JsonObject {
  [key: string]: any;
}

type RequestMethod = "GET" | "POST";

const JSON_CONTENT = "application/json";
export const headers = {
  "Content-Type": JSON_CONTENT,
  Accept: JSON_CONTENT,
};

/**
 * Helper function to handle paths with or with out leading slash
 * example:
 * joinAbsoluteUrlPath("a/b", "/c/d/", "/e", "f/g", "h") -> "/a/b/c/d/e/f/g/h"
 */
function joinAbsoluteUrlPath(...args: string[]) {
  return (
    "/" + args.map((pathPart) => pathPart.replace(/(^\/|\/$)/g, "")).join("/")
  );
}
class FetchClient {
  basePath: string;

  constructor(basePath: string) {
    this.basePath = "/api" + basePath;
  }

  getURL = (path: string, query: string) => {
    if (!process.env.REACT_APP_BACKEND_URL) {
      throw Error("process.env.REACT_APP_BACKEND_URL is falsy");
    }
    return new URL(
      joinAbsoluteUrlPath(this.basePath, path) + query,
      process.env.REACT_APP_BACKEND_URL
    ).href;
  };

  getOptions = (
    method: RequestMethod,
    body: JsonObject | null
  ): RequestInit => {
    return {
      method,
      headers,
      body: body ? JSON.stringify(body) : undefined,
    };
  };

  request = async (
    path: string,
    body: JsonObject | null = null,
    method: RequestMethod = "POST",
    query = ""
  ) => {
    const res = await fetch(
      this.getURL(path, query),
      this.getOptions(method, body)
    );
    if (!res.ok) {
      throw await res.json();
    }
    const contentType = res.headers.get("content-type");
    if (contentType && contentType.indexOf(JSON_CONTENT) !== -1) {
      try {
        return await res.json();
      } catch {
        throw new Error("Invalid JSON response");
      }
    } else {
      return "success";
    }
  };
}

export default FetchClient;
