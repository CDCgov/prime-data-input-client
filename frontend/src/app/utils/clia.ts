import { states, noCLIAValidationStates } from "../../config/constants";

export function stateRequiresCLIANumberValidation(
  state: keyof typeof states
): boolean {
  return !noCLIAValidationStates.includes(state);
}

export function isValidCLIANumber(input: string): boolean {
  const cliaNumberValidator = /^\d{2}D\d{7}$/;

  return cliaNumberValidator.test(input);
}
