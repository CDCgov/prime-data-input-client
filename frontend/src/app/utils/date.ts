import moment from "moment";

export const daysSince = (date: moment.Moment): String => {
  const now = moment();
  const days = now.diff(date, "days");
  return `${days} day${days === 1 ? "" : "s"}`;
};
