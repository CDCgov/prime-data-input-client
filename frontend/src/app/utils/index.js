import Alert from "../commonComponents/Alert";
import React from "react";

export const displayFullName = (first, middle, last) => {
  return `${first || "?"} ${middle || ""} ${last || "?"}`.replace(/ +/g, " ");
};

export const isLocalHost = () =>
  Boolean(window.location.hostname === "localhost");

export const showNotification = (toast, children) => {
  try {
    // id will de-dup. just use whole message as id
    const toastId = JSON.stringify(children.props).substr(0, 512);
    toast(children, { toastId });
    toast.clearWaitingQueue(); // don't pile up messages
  } catch (err) {
    console.error(err, err.stack);
  }
};

export const showError = (
  toast,
  message = "Please check for missing data or typos.",
  title = "Problems saving data to server"
) => {
  const err_msg = message.substr(0, 512);
  showNotification(toast, <Alert type="error" title={title} body={err_msg} />);
};
