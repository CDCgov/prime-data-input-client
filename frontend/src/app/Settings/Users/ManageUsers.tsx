import React, { useEffect, useState } from "react";
import { Prompt } from "react-router-dom";
import { toast } from "react-toastify";

import Alert from "../../commonComponents/Alert";
import Button from "../../commonComponents/Button";
import CreateUserModal from "./CreateUserModal";
import DeleteUserModal from "./DeleteUserModal";
import InProgressModal from "./InProgressModal";
import UserFacilitiesSettingsForm from "./UserFacilitiesSettingsForm";
import UserRoleSettingsForm from "./UserRoleSettingsForm";
import UsersSideNav from "./UsersSideNav";
import {
  SettingsUser,
  UserFacilitySetting,
  NewUserInvite,
} from "./ManageUsersContainer";
import {
  showNotification,
  displayFullNameInOrder,
  displayFullName,
} from "../../utils";
import { OrganizationRole, RoleDescription } from "../../permissions";

import "./ManageUsers.scss";

const RoleDescriptionToOrgRole = {
  "Admin user": "ADMIN",
  "Standard user": "USER",
  "Test-entry user": "ENTRY_ONLY",
};

interface Props {
  users: SettingsUser[];
  loggedInUser: User;
  allFacilities: UserFacilitySetting[];
  updateUserRole: (variables: any) => Promise<any>;
  addUserToOrg: (variables: any) => Promise<any>;
  deleteUser: (variables: any) => Promise<any>;
  getUsers: () => Promise<any>;
}

export type SettingsUsers = { [id: string]: SettingsUser };

const sortUsers = (users: SettingsUsers) =>
  Object.values(users).sort((a, b) => {
    const nameA = displayFullName(a.firstName, a.middleName, a.lastName);
    const nameB = displayFullName(b.firstName, b.middleName, b.lastName);
    if (nameA === nameB) return 0;
    return nameA > nameB ? 1 : -1;
  });

const getSettingsUser = (users: SettingsUser[]) =>
  users.reduce((acc: SettingsUsers, user: SettingsUser) => {
    acc[user.id] = user;
    return acc;
  }, {});

const ManageUsers: React.FC<Props> = ({
  users,
  loggedInUser,
  allFacilities,
  updateUserRole,
  addUserToOrg,
  deleteUser,
  getUsers,
}) => {
  const [activeUser, updateActiveUser] = useState<SettingsUser>();
  const [nextActiveUserId, updateNextActiveUserId] = useState<string | null>(
    null
  );
  const [showInProgressModal, updateShowInProgressModal] = useState(false);
  const [showAddUserModal, updateShowAddUserModal] = useState(false);
  const [showDeleteUserModal, updateShowDeleteUserModal] = useState(false);
  const [isUserEdited, updateIsUserEdited] = useState(false);
  const [isUpdating, setIsUpdating] = useState(false);
  const [error, setError] = useState<Error>();
  // Maintain deleted user ID on client while request is in flight
  const [deletedUserId, setDeletedUserId] = useState<string>();
  // Maintain added user on client while request is in flight
  const [addedUserId, setAddedUserId] = useState<string>();

  if (error) {
    throw error;
  }

  // Local users will be more up-to-date than users, which only gets updated after server requests complete
  let localUsers = [...users];
  if (deletedUserId) {
    localUsers = localUsers.filter(({ id }) => id !== deletedUserId);
  }
  const usersState: SettingsUsers = getSettingsUser(localUsers);
  const sortedUsers = sortUsers(usersState);

  // only updates the local state
  function updateUser<T>(
    userId: string,
    key: string, // the field to update
    value: T // value of the field to update
  ) {
    updateActiveUser({
      ...usersState[userId],
      [key]: value,
    });
    updateIsUserEdited(true);
  }

  // confirm with the user if they have unsaved edits and want to change users
  const onChangeActiveUser = (nextActiveUserId: string) => {
    if (isUserEdited) {
      updateNextActiveUserId(nextActiveUserId);
      updateShowInProgressModal(true);
    } else {
      updateActiveUser(usersState[nextActiveUserId]);
    }
  };

  // proceed with changing the active user
  const onContinueChangeActiveUser = () => {
    updateIsUserEdited(false);
    updateShowInProgressModal(false);
    if (nextActiveUserId) {
      updateActiveUser(usersState[nextActiveUserId]);
    }
  };

  const handleUpdateUser = () => {
    if (!activeUser) {
      return;
    }
    setIsUpdating(true);
    const selectedRoleDescription = activeUser.roleDescription as RoleDescription;
    const selectedOrganizationRole = RoleDescriptionToOrgRole[
      selectedRoleDescription
    ] as OrganizationRole;
    updateUserRole({
      variables: {
        id: activeUser.id,
        role: selectedOrganizationRole,
      },
    })
      .then(() => {
        getUsers();
        updateIsUserEdited(false);
        const user = usersState[activeUser.id];
        const fullName = displayFullNameInOrder(
          user.firstName,
          user.middleName,
          user.lastName
        );

        setIsUpdating(false);
        showNotification(
          toast,
          <Alert
            type="success"
            title="Changes Saved"
            body={`${fullName}'s settings have been saved`}
          />
        );
      })
      .catch(setError);
  };

  const handleAddUserToOrg = async (newUserInvite: NewUserInvite) => {
    // TODO: validate form
    try {
      setIsUpdating(true);
      const { firstName, lastName, email, role } = { ...newUserInvite };
      const { data } = await addUserToOrg({
        variables: { firstName, lastName, email, role: role || undefined },
      });
      await getUsers();
      const fullName = displayFullNameInOrder(firstName, "", lastName);
      showNotification(
        toast,
        <Alert
          type="success"
          title={`Invitation sent to ${fullName}`}
          body={`They will receive an invitation to create an account at the email address provided`}
        />
      );
      updateShowAddUserModal(false);
      setAddedUserId(data?.addUserToCurrentOrg?.id);
      setIsUpdating(false);
    } catch (e) {
      setError(e);
    }
  };

  const handleDeleteUser = async (userId: string) => {
    try {
      const data = await deleteUser({
        variables: {
          id: userId,
          deleted: true,
        },
      });
      const deletedUserId = data.data.setUserIsDeleted.id;
      const user = usersState[deletedUserId];

      const fullName = displayFullNameInOrder(
        user.firstName,
        user.middleName,
        user.lastName
      );
      updateShowDeleteUserModal(false);
      setDeletedUserId(userId);
      showNotification(
        toast,
        <Alert type="success" title={`User account removed for ${fullName}`} />
      );
      await getUsers();
      updateActiveUser(sortedUsers[0]); // arbitrarily pick the first user as the next active.
      setDeletedUserId(undefined);
    } catch (e) {
      setError(e);
    }
  };

  // Default to first user
  useEffect(() => {
    if (!activeUser && sortedUsers.length) {
      updateActiveUser(sortedUsers[0]);
    }
  }, [activeUser, sortedUsers]);

  // Navigate to addeed user, if applicable
  useEffect(() => {
    if (addedUserId && usersState[addedUserId]) {
      updateActiveUser(usersState[addedUserId]);
      setAddedUserId(undefined);
    }
  }, [addedUserId, usersState]);

  return (
    <div className="prime-container usa-card__container">
      <div className="usa-card__header">
        <h2>Manage Users</h2>
        {process.env.REACT_APP_ADD_NEW_USER_ENABLED === "true" ? (
          <Button
            variant="outline"
            onClick={() => updateShowAddUserModal(true)}
            label="+ New user"
          />
        ) : null}
      </div>
      {!activeUser || !localUsers.length ? (
        <div className="usa-card__body">
          {!localUsers.length ? (
            <p>There are no users in this organization</p>
          ) : (
            <p>Loading user data</p>
          )}
        </div>
      ) : (
        <div className="usa-card__body">
          <div className="grid-row">
            <UsersSideNav
              activeUserId={activeUser.id || ""}
              users={sortedUsers}
              onChangeActiveUser={onChangeActiveUser}
            />
            <div className="tablet:grid-col padding-left-2">
              <div className="user-header grid-row flex-row flex-align-center">
                <h2 className="display-inline-block margin-y-1">
                  {displayFullNameInOrder(
                    activeUser.firstName,
                    activeUser.middleName,
                    activeUser.lastName
                  )}
                </h2>
                {activeUser?.id === loggedInUser.id ? (
                  <span className="usa-tag margin-left-1 bg-base-lighter text-ink">
                    YOU
                  </span>
                ) : null}
              </div>
              <div className="user-content">
                <p className="text-base">
                  Admins have full access to conduct tests, manage results and
                  profiles, and manage settings and users
                </p>
                {
                  <UserRoleSettingsForm
                    activeUser={activeUser}
                    loggedInUser={loggedInUser}
                    onUpdateUser={updateUser}
                  />
                }

                {process.env.REACT_APP_VIEW_USER_FACILITIES === "true" ? (
                  <UserFacilitiesSettingsForm
                    activeUser={activeUser}
                    allFacilities={allFacilities}
                    onUpdateUser={updateUser}
                  />
                ) : null}
              </div>
              <div className="usa-card__footer display-flex flex-justify margin-top-5 padding-x-0">
                {process.env.REACT_APP_DELETE_USER_ENABLED === "true" ? (
                  <Button
                    variant="outline"
                    icon="trash"
                    className="flex-align-self-start display-inline-block"
                    onClick={() => updateShowDeleteUserModal(true)}
                    label="Remove user"
                    disabled={loggedInUser.id === activeUser.id || isUpdating}
                  />
                ) : null}
                <Button
                  type="button"
                  onClick={() => handleUpdateUser()}
                  label={isUpdating ? "Saving..." : "Save changes"}
                  disabled={
                    // enabled only if the user has been edited AND the loggedInUser is an org admin or super admin
                    !isUserEdited ||
                    !["Admin user", "Admin user (SU)"].includes(
                      loggedInUser.roleDescription
                    ) ||
                    isUpdating
                  }
                />
              </div>

              {showInProgressModal && (
                <InProgressModal
                  onClose={() => updateShowInProgressModal(false)}
                  onContinue={() => onContinueChangeActiveUser()}
                />
              )}
              {isUserEdited ? (
                <Prompt
                  when={isUserEdited}
                  message="You have unsaved changes. Do you want to continue?"
                />
              ) : null}
              {showAddUserModal &&
              process.env.REACT_APP_ADD_NEW_USER_ENABLED === "true" ? (
                <CreateUserModal
                  isUpdating={isUpdating}
                  onClose={() => updateShowAddUserModal(false)}
                  onSubmit={handleAddUserToOrg}
                />
              ) : null}
              {showDeleteUserModal &&
              process.env.REACT_APP_DELETE_USER_ENABLED === "true" ? (
                <DeleteUserModal
                  user={activeUser}
                  onClose={() => updateShowDeleteUserModal(false)}
                  onDeleteUser={handleDeleteUser}
                />
              ) : null}
            </div>
          </div>
        </div>
      )}
    </div>
  );
};

export default ManageUsers;
