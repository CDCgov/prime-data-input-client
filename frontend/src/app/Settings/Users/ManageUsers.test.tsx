import { render, fireEvent, waitFor } from "@testing-library/react";
import { displayFullNameInOrder } from "../../utils";
import ManageUsers, { SettingsUsers } from "./ManageUsers";

const organization = { testingFacility: [{ id: "a1", name: "Foo Org" }] };
const allFacilities = [
  organization.testingFacility[0],
  { id: "a2", name: "Bar Org" },
];

const loggedInUser = {
  firstName: "Bob",
  middleName: "",
  lastName: "Bobberoo",
  id: "b1",
  email: "bob@bobberoo.org",
  suffix: "",
  roleDescription: "Admin user",
};

const users: SettingsUsers[keyof SettingsUsers][] = [
  {
    firstName: "John",
    middleName: "",
    lastName: "Doe",
    id: "a123",
    email: "john@doe.org",
    organization,
    permissions: ["READ_PATIENT_LIST"],
    roleDescription: "user",
  },
  { ...loggedInUser, permissions: [], organization, roleDescription: "admin" },
];

let updateUserRole: () => Promise<any>;
let addUserToOrg: () => Promise<any>;
let deleteUser: () => Promise<any>;
let getUsers: () => void;

let inputValue = (value: string) => ({ target: { value } });

describe("ManageUsers", () => {
  beforeEach(() => {
    updateUserRole = jest.fn(() => Promise.resolve());
    addUserToOrg = jest.fn(() => Promise.resolve());
    deleteUser = jest.fn(() => Promise.resolve());
    getUsers = jest.fn();
  });
  it("displays the list of users and defaults to the first user", () => {
    const { container } = render(
      <ManageUsers
        users={users}
        loggedInUser={loggedInUser}
        allFacilities={allFacilities}
        updateUserRole={updateUserRole}
        addUserToOrg={addUserToOrg}
        deleteUser={deleteUser}
        getUsers={getUsers}
      />
    );
    expect(container).toMatchSnapshot();
  });
  it("disables logged-in user's settings", async () => {
    const { container, findByText, getByText } = render(
      <ManageUsers
        users={users}
        loggedInUser={loggedInUser}
        allFacilities={allFacilities}
        updateUserRole={updateUserRole}
        addUserToOrg={addUserToOrg}
        deleteUser={deleteUser}
        getUsers={getUsers}
      />
    );

    fireEvent.click(getByText(displayFullNameInOrder("Bob", "", "Bobberoo")));
    await findByText("YOU");
    expect(container).toMatchSnapshot();
  });
  it("passes user details to the addUserToOrg function", async () => {
    const { getByText, findAllByRole } = render(
      <ManageUsers
        users={users}
        loggedInUser={loggedInUser}
        allFacilities={allFacilities}
        updateUserRole={updateUserRole}
        addUserToOrg={addUserToOrg}
        deleteUser={deleteUser}
        getUsers={getUsers}
      />
    );

    const newUser = {
      firstName: "Jane",
      lastName: "Smith",
      email: "jane@smith.co",
    };

    fireEvent.click(getByText("New User", { exact: false }));
    const [first, last, email] = await findAllByRole("textbox");
    fireEvent.change(first, inputValue(newUser.firstName));
    fireEvent.change(last, inputValue(newUser.lastName));
    fireEvent.change(email, inputValue(newUser.email));
    fireEvent.click(getByText("Send", { exact: false }));
    await waitFor(() => expect(addUserToOrg).toBeCalled());
    expect(addUserToOrg).toBeCalledWith({ variables: newUser });
  });
});
