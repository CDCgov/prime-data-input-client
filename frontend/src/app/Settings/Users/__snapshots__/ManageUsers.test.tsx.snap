// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageUsers displays the list of users and defaults to the first user 1`] = `
<div>
  <div
    class="prime-container card-container"
  >
    <div
      class="usa-card__header"
    >
      <h2>
        Manage Users
      </h2>
      <button
        class="usa-button usa-button--outline"
        type="button"
      >
        + New user
      </button>
    </div>
    <div
      class="usa-card__body"
    >
      <p>
        Loading user data
      </p>
    </div>
  </div>
</div>
`;
